### ISSUE TITLE
Firewall support

### ISSUE THREAD

**Issue by tristanz** [kubernetes/kubernetes#2880]
As Kubernetes moves to support more multi-tenant use cases and to increase security in general, it would be helpful if there was a way to firewall pods from other pods, ideally using something simple like labels.  

Is there a preferred approach to support this use case?  If this is implemented via dynamic iptables rules, I assume it would need to be part of the pod definition and kubelet rather than an external firewall service? Or perhaps there are lessons learned at Google regarding how to best isolate applications?

Somewhat related #2585.

**Comment by Symmetric** [comment]
I've implemented a prototype of pod firewalling, using the [Calico](https://github.com/Metaswitch/calico-docker) networking framework. It's on a branch in my repo [here](https://github.com/Metaswitch/calico-kubernetes/tree/calico-with-services).

I added annotation directly to the ports dictionary in the PodSpec; that's just for clarity. So with a spec like this:

```
apiVersion: v1beta1
desiredState:
  manifest:
    containers:
      - image: busybox
        name: demo-busybox
        command:
          - "/bin/sh"
          - "-c"
          - "while sleep 10; do date; done"
        ports: 
          - containerPort: 80
            name: http
            allowFrom: 
              app: db
    id: demo-busybox
    version: v1beta1
labels:
  app: sleep
  stage: prod
id: demo-busybox
kind: Pod
```

any pod with label `app: db` would be able to access port 80 on the demo-busybox pod. Traffic from pods with label `app: db` to other ports will be firewalled, as will all traffic from pods without that label.

Thoughts on the syntax? The allowFrom field could also be an annotation, but I think there's an advantage to making this a first-class citizen in the API, since it reduces duplication (with an annotation you'd have to specify the port in two places).

This could easily be extended to include `allowTo` as well, so that traffic is firewalled at both ends.

Note that this currently only works with headless services, since for normal services the kube-proxy prevents destination filtering (since all traffic at the destination pod has the src_ip of the proxy, and therefore we can't determine the original sender).

**Comment by thockin** [comment]
We need to figure out the semantics before the syntax.  The simplest form
is per-namespace firewalls, which I agree is probably the direction to go.
We'll have to address cluster-services, like DNS, as part of this.

Then we have to decide if we want to really go to sub-namespace subnets at
all, and if so, how do we denote that such that network environments that
don't support it work properly (where "properly" is TBD)

**Comment by Symmetric** [comment]
Had a wee think about this recently, here are a few thoughts/questions that come to mind on the subject of namespace firewalling:

1. If we're firewalling namespaces, we will obviously need to be able to allow traffic from outside the cluster.  Service.Type=LoadBalancer still gets us that.
2. To allow "global services", i.e. accessible from any namespace in the cluster but not externally, Service.Type.NodePort should work. Is this the right way to achieve this? Is there/do we need a different mechanism?
3. Do we want to be able to prevent pods in namespace foo from accessing externally exposed services in namespace bar? This would make things more complicated, not sure if there's a strong use-case here.
4. Conversely to 3), do we want to be able to allow traffic between namespaces foo and bar, without using the "global service" construct? (The use-case would be something like a 'backend' namespace, which might contain a bunch of namespace-private pods, and expose a single service to be used by pods in the 'frontend' namespace -- but not by any other namespaces. For distinct and sufficiently large 'backend' and 'frontend' teams, this sort of policing might be desirable. Perhaps this is a perversion of the namespace semantics though.)
5. As an alternative to 3) and 4), if we want access control, we can implement fine-grained intent-based ACL on Pods, and omit the complexity of specifying connectivity between namespaces. (As examplified in my previous comment above; I like this more than 3 & 4). Note this isn't mutually exclusive with 3) & 4), some sort of layered firewalling could be useful to provide defense in depth, i.e. fine-grained policy at the pod level as the primary isolation mechanism, with namespace-level coarse-grained policy as a backstop to impose organizational constraints.

There are many layers to this onion. It feels like the minimal viable increment would be 1) & 2), giving completely firewalled namespaces, "global" cluster services with NodePort, and external cluster services using LoadBalancer.

As for implementation of any such firewalling, is this something that gets pulled into the (new, iptables-based) kube-proxy, is it strictly delegated to network plugins, or is it a "batteries included but removable" style feature?

cc @lxpollitt

**Comment by tristanz** [comment]
1) and 2) would be a good fit for our use cases:  completely firewalled namespaces, cluster services, and external services.  

I can see the value of 4) and 5) but I'm not sure I should even be aware of other namespaces and definitely not pod labels in other namespaces so writing rules based on these seems off.

**Comment by Symmetric** [comment]
cc @spikecurtis

**Comment by lxpollitt** [comment]
I agree with @tistanz that a pod in one namespace probably shouldn't have any visibility to the labels of pods in another name space.  I do think there is value in pods within a namespace being able to define policy for how they want to interact with other pods in the namespace and the outside world though.

**Comment by thockin** [comment]
On Fri, May 29, 2015 at 10:37 AM, Paul Tiplady <notifications@github.com> wrote:
>
> Had a wee think about this recently, here are a few thoughts/questions that come to mind on the subject of namespace firewalling:
>
> If we're firewalling namespaces, we will obviously need to be able to allow traffic from outside the cluster. Service.Type=LoadBalancer still gets us that.

external load balancers present an interesting challenge wrt
firewalling because they do not really exist in the kubernetes space,
but we need to poke holes for them.

> To allow "global services", i.e. accessible from any namespace in the cluster but not externally, Service.Type.NodePort should work. Is this the right way to achieve this? Is there/do we need a different mechanism?

NodePort is not the right mechanism.  NodePort is, first and foremost,
a way of publishing to outside the cluster.  We'll have to decide just
how draconian we intend to be wrt services across namespaces - it
probably should include locking down node ports.

> Do we want to be able to prevent pods in namespace foo from accessing externally exposed services in namespace bar? This would make things more complicated, not sure if there's a strong use-case here.

I think that, from a utilitarian POV, once you publish without
restrictions to a scope higher than your own, we stop trying to lock
it down.  If you publish a service to all namespaces, we don't try to
restrict.  If you publish to the internet, we don't try to restrict.

> Conversely to 3), do we want to be able to allow traffic between namespaces foo and bar, without using the "global service" construct? (The use-case would be something like a 'backend' namespace, which might contain a bunch of namespace-private pods, and expose a single service to be used by pods in the 'frontend' namespace -- but not by any other namespaces. For distinct and sufficiently large 'backend' and 'frontend' teams, this sort of policing might be desirable. Perhaps this is a perversion of the namespace semantics though.)

I was discussing this with @jackgr just last week.  I think it's a
reasonable concept to export/import services across namespaces, but I
have not put a lot of thought into ACLing that.

> As an alternative to 3) and 4), if we want access control, we can implement fine-grained intent-based ACL on Pods, and omit the complexity of specifying connectivity between namespaces. (As examplified in my previous comment above; I like this more than 3 & 4). Note this isn't mutually exclusive with 3) & 4), some sort of layered firewalling could be useful to provide defense in depth, i.e. fine-grained policy at the pod level as the primary isolation mechanism, with namespace-level coarse-grained policy as a backstop to impose organizational constraints.
>
> There are many layers to this onion. It feels like the minimal viable increment would be 1) & 2), giving completely firewalled namespaces, "global" cluster services with NodePort, and external cluster services using LoadBalancer.

I agree that segmenting on namespace boundaries is obvious and useful.
I would like to see some comparative analysis of different approaches
to cross-ns or global services.  e.g.

1. flag namespace as "open" to all
2. flag namespace as "open" to some but not all
3. flag services as exported, import into consumers
4. publish services up to a higher-level or parallel scope
5. ...?

> As for implementation of any such firewalling, is this something that gets pulled into the (new, iptables-based) kube-proxy, is it strictly delegated to network plugins, or is it a "batteries included but removable" style feature?

ideally it's pluggable, but perhaps it doesn't start there.

**Comment by crizCraig** [comment]
Looks like Calico has a plugin for this: https://www.projectcalico.org/calico-network-policy-comes-to-kubernetes/

**Comment by caseydavenport** [comment]
I believe this is covered by https://github.com/kubernetes/kubernetes/pull/25638 and related PRs.  NetworkPolicy is available in v1.3.0 for network plugins which support it (e.g Calico).

I think this issue can be closed.
